FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -0,0 +1,207 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import FirebaseCore
+import FirebaseInstallations
+
+/// Model info object with details about pending or downloaded model.
+struct ModelInfo {
+  /// Model name.
+  var name: String
+
+  /// User defaults associated with model.
+  var defaults: UserDefaults
+
+  /// Download URL for the model file, as returned by server.
+  @UserDefaultsBacked var downloadURL: String
+
+  /// Hash of the model, as returned by server.
+  @UserDefaultsBacked var hash: String
+
+  /// Size of the model, as returned by server.
+  @UserDefaultsBacked var size: Int
+
+  /// Local path of the model.
+  @UserDefaultsBacked var path: String?
+
+  /// Initialize model info and create user default keys.
+  init(app: FirebaseApp, name: String, defaults: UserDefaults = .firebaseMLDefaults) {
+    self.name = name
+    self.defaults = defaults
+    let bundleID = Bundle.main.bundleIdentifier ?? ""
+    let defaultsPrefix = "\(bundleID).\(app.name).\(name)"
+    _downloadURL = UserDefaultsBacked(
+      key: "\(defaultsPrefix).model-download-url",
+      storage: defaults
+    )
+    _hash = UserDefaultsBacked(key: "\(defaultsPrefix).model-hash", storage: defaults)
+    _size = UserDefaultsBacked(key: "\(defaultsPrefix).model-size", storage: defaults)
+    _path = UserDefaultsBacked(key: "\(defaultsPrefix).model-path", storage: defaults)
+  }
+}
+
+/// Model info retriever for a model from local user defaults or server.
+class ModelInfoRetriever: NSObject {
+  /// Current Firebase app.
+  var app: FirebaseApp
+  /// Model info associated with model.
+  var modelInfo: ModelInfo?
+  /// Project id.
+  var projectID: String
+  /// Model name.
+  var modelName: String
+  /// Firebase installations.
+  var installations: Installations
+
+  /// Associate model info retriever with current Firebase app, project ID, and model name.
+  init(app: FirebaseApp, projectID: String, modelName: String) {
+    self.app = app
+    self.projectID = projectID
+    self.modelName = modelName
+    installations = Installations.installations(app: app)
+  }
+
+  /// Build custom model object from model info.
+  func buildModel() -> CustomModel? {
+    /// Build custom model only if model info is filled out, and model file is already on device.
+    guard let info = modelInfo, let path = info.path else { return nil }
+    let model = CustomModel(
+      name: info.name,
+      size: info.size,
+      path: path,
+      hash: info.hash
+    )
+    return model
+  }
+}
+
+/// Extension to handle fetching model info from server.
+extension ModelInfoRetriever {
+  /// HTTP request headers.
+  static let fisTokenHTTPHeader: String = "x-goog-firebase-installations-auth"
```

Comment:
nit: the `String` type declaration is redundant when assigned with the literal here and below. It should be just:
```
static let fisTokenHTTPHeader = "x-goog-firebase-installations-auth"
```

Reply:
Done.